parley/src/layout/cursor.rs

```diff
@@ -155,6 +155,44 @@ impl Cursor {
     pub fn is_trailing(&self) -> bool {
         self.text_end == self.insert_point
     }
+
+    /// Given the layout that generated this cursor, return a new cursor
+    /// for the corresponding position on the next line.
+    ///
+    /// If `h_pos` is provided, then it will be used as the horizontal offset
+    /// for computing the position on the next line.
+    ///
+    /// Returns `None` if the cursor should remain in its current position.
+    pub fn next_line<B: Brush>(&self, layout: &Layout<B>, h_pos: Option<f32>) -> Option<Cursor> {
+        move_to_line(layout, self, h_pos, self.path.line_index.checked_add(1)?)
+    }
+
+    /// Given the layout that generated this cursor, return a new cursor
+    /// for the corresponding position on the previous line.
+    ///
+    /// If `h_pos` is provided, then it will be used as the horizontal offset
+    /// for computing the position on the previous line.
+    ///
+    /// Returns `None` if the cursor should remain in its current position.
+    pub fn prev_line<B: Brush>(&self, layout: &Layout<B>, h_pos: Option<f32>) -> Option<Cursor> {
+        move_to_line(layout, self, h_pos, self.path.line_index.checked_sub(1)?)
+    }
+}
+
+fn move_to_line<B: Brush>(
+    layout: &Layout<B>,
+    cursor: &Cursor,
+    h_pos: Option<f32>,
+    line_index: usize,
+) -> Option<Cursor> {
+    let line = layout.get(line_index)?;
+    let metrics = line.metrics();
+    let y = metrics.baseline - metrics.line_height * 0.5;
+    Some(Cursor::from_point(
+        layout,
+        h_pos.unwrap_or(cursor.offset),
+        y,
+    ))
```

Comment:
A bit silly to be working out the y pos from the line and then `Cursor::from_point` presumably turning that back into a line, but it'll certainly work.

Reply:
Agreed. In theory, I'd rather see it use the data we already know will just be re-found to be the index we have available. But this is fine as-is.

Reply:
I disagree. This stuff is really complex (the current code in masonry gets much of it wrong and will be worse with mixed direction text). Once you have fast and robust index/point primitives, you want to use them as much as possible.

Reply:
Line movement is fundamentally a hit test operation anyway, given the expectation of maintaining the visual horizontal position.

Reply:
I mean sure. It would be an optimisation to skip hit testing in the y direction because you already know which line you will hit (right?). If you think it's better not to do that to keep the code simpler then that seems fine. It's not something I feel particularly strongly about (and would also be something that would be easy to change later (perhaps when the codebase is more mature and better tested))

Reply:
Point taken, we could factor the horizontal search out of `from_position` and reuse it for this.